index.js

```diff
@@ -14,44 +13,89 @@ function createResolver(config, options) {
     resolve: resolve,
   };
 
-  var lastKey;
 
-  function resolve(key) {
-    if (typeof key !== 'string') {
+  // Keep constants separately
+  var constants = {};
+
+  function resolveConstant(key) {
+    if (constants.hasOwnProperty(key)) {
+      return constants[key];
+    }
+
+    var definition = config[key];
+    // Ignore options that are not defined
+    if (!definition) {
       return;
     }
 
-    var appliedArgs = slice.call(arguments, 1);
+    var option = options[key];
+
+    if (!!option || options.hasOwnProperty(key)) {
+      if (typeof option !== 'function') {
+        option = normalize.call(resolver, definition.type, option);
+        if (option != null) {
+          constants[key] = option;
+          return option;
+        }
+        // Fall through
+      } else {
+        return;
+      }
+    }
+
+    var fallback = definition.default;
+    if (option == null && typeof fallback !== 'function') {
+      constants[key] = fallback;
+      return fallback;
+    }
+  }
+
+
+  // Keep requested keys to detect (and disallow) recursive resolution
+  var stack = [];
+
+  function resolve(key) {
+    var option = resolveConstant(key);
+    if (option != null) {
+      return option;
+    }
 
     var definition = config[key];
     // Ignore options that are not defined
     if (!definition) {
       return;
     }
 
-    if (key === lastKey) {
+    if (stack.indexOf(key) >= 0) {
       throw new Error('Recursive resolution denied.');
     }
-    lastKey = key;
-
-    var option = options[key];
-    // Bind the option so it can resolve other options if necessary
-    if (typeof option === 'function') {
-      option = option.bind(resolver);
-    }
 
+    option = options[key];
+    var fallback = definition.default;
+    var appliedArgs = slice.call(arguments, 1);
     var args = [definition.type, option].concat(appliedArgs);
-    var result = normalize.apply(null, args);
 
-    var fallback = definition.default;
-    // Bind & apply the default so it can resolve other options if necessary
-    if (typeof fallback === 'function') {
-      fallback = fallback.apply(resolver, appliedArgs);
-    }
+    stack.push(key);
+    return tryResolve(option, fallback, appliedArgs, args);
+  }
 
-    lastKey = null;
 
-    return koalas(result, fallback);
+  function tryResolve(option, fallback, appliedArgs, args) {
+    try {
+      option = normalize.apply(resolver, args);
```

Comment:
I think I just did what you requested, but actually I believe the performance hit of try/catch was already averted? Still, this newer version has to pass around some fewer args, so that's probably for the best.